Assets/Shopify/UIToolkit/Components/RemoteImageLoader.cs

```diff
@@ -0,0 +1,140 @@
+            _www = new WWW(url, null, headers);
+            yield return _www;
+        }
+    }
+
+    /// <summary>
+    /// A behaviour that fetches and caches a remote image (JPEG/PNG) from the web.!-- from the web.
+    /// </summary>
+    public class RemoteImageLoader : MonoBehaviour {
+
+        private class HTTPHeaderFields {
+            public const string LAST_MODIFIED = "Last-Modified";
+            public const string STATUS_CODE = "STATUS";
+            public const string IF_MODIFIED_SINCE = "If-Modified-Since";
+        }
+
+        /// <summary>
+        /// The resulting error after downloading a web image.
+        /// </summary>
+        /// <returns></returns>
+        public string error {
+            get {
+                return _error;
+            }
+        }
+
+        /// <summary>
+        /// The resulting Texture2D web image after downloading from the given URL.
+        /// </summary>
+        /// <returns></returns>
+        public Texture2D texture {
+            get {
+                return _texture;
+            }
+        }
+
+        private const int HTTP_STATUS_NOT_MODIFIED = 304;
+
+        private WebImageCache _imageCache;
+        private string _error;
+        private Texture2D _texture;
+        private ImageLoader _loader;
+
+        public void SetImageLoader(ImageLoader loader) {
+            _loader = loader;
+        }
+
+        /// <summary>
+        /// Downloads the given web image located at the URL.
+        /// </summary>
+        /// <param name="url">URL of the image resource to download and cache.</param>
+        /// <param name="cache">Determine if the downloaded resource should be cached or not.</param>
+        /// <returns>On completion, check texture/error properties for result.</returns>
+        public IEnumerator DownloadImageURL(string url, bool cache = true) {
+            _loader = _loader != null ? _loader : new UnityImageLoader();
+            var imageCache = WebImageCache.SharedCache;
+
+            // If we have a cached texture for this URL, grab it's Last-Modified timestamp to see if we need to invalidate.
+            var requestHeaders = new Dictionary<string, string>();
+            CachedWebResource<Texture2D>? cachedResource = imageCache.TextureResourceForURL(url);
+            if (cachedResource != null) {
+                requestHeaders[HTTPHeaderFields.IF_MODIFIED_SINCE] = cachedResource.Value.LastModifiedTimestamp;
+            }
+
+            yield return _loader.Load(url, requestHeaders);
+
+            // Bail out early if we hit an error.
+            if (_loader.GetError() != null) {
+                _error = _loader.GetError();
+                yield return null;
+            }
+
+            var responseHeaders = _loader.GetResponseHeaders();
+
+            // Check to see if we got a 304 Not Modified response. In that case, just return the cached image.
+            if (responseHeaders.ContainsKey(HTTPHeaderFields.STATUS_CODE)) {
```

Comment:
If we have to make the request anyways, why not just use the downloaded image? The damage is already done? Unless I'm missing something?

Reply:
In most cases we can probably get by with just using the saved texture directly. Using the Last-Modified/If-Not-Modified dance gives us the ability to send a super cheap partial GET and only download the image if it has changed. For our in-memory cache we probably don't need this complexity but we could cache the images on disk using this method without worrying about them getting stale. Might be nice to implement a disk cache for these to make the store images load instantly on subsequent launches of the game.